src/codegen/wasm.zig

```diff
@@ -305,53 +314,54 @@ pub const Context = struct {
         return WValue{ .local = self.local_index };
     }
 
-    fn genAdd(self: *Context, inst: *Inst.BinOp) InnerError!WValue {
+    fn genBinOp(self: *Context, inst: *Inst.BinOp, op: wasm.Op) InnerError!WValue {
         const lhs = self.resolveInst(inst.lhs);
         const rhs = self.resolveInst(inst.rhs);
 
         try self.emitWValue(lhs);
         try self.emitWValue(rhs);
 
-        const opcode: wasm.Opcode = switch (inst.base.ty.tag()) {
-            .u32, .i32 => .i32_add,
-            .u64, .i64 => .i64_add,
-            .f32 => .f32_add,
-            .f64 => .f64_add,
-            else => return self.fail(inst.base.src, "TODO - Implement wasm genAdd for type '{s}'", .{inst.base.ty.tag()}),
-        };
-
-        try self.code.append(wasm.opcode(opcode));
-        return .none;
+        switch (inst.base.ty.tag()) {
+            .u32, .i32, .u64, .i64, .f32, .f64 => {
```

Comment:
Not really a fan of this approach. This effectively means we'll have to maintain all the switch branches here, as well as in `typeToValtype`. Instead, you should let `typeToValtype` do the work and generate an error if some unhandled/invalid type is caught. That way it becomes more maintainable as well as readable *imo*.

Reply:
This was based on an incorrect understanding of the compiler. BinOp always means numbers.

Since this only does numbers, we should always emit the same wasm (only differing in the valtype: `[iuf](32|64)`), so `typeToValtype` should handle all cases correctly.

Reply:
Good point! Perhaps we should be switching on a different type/tag instead? I.e. switching on `zigTypeTag()` would allow us to switch on `.Int`, `.Pointer` etc. and means we don't have to be as specific in regards to the tags for now.

Reply:
I think in this case, we can panic if `zigTypeTag()` is neither `.Int`, `.Float`, nor `.Bool`. the only `BinOp`s are:

```rs
                .add,
                .addwrap,
                .sub,
                .subwrap,
                .mul,
                .mulwrap,
                .cmp_lt,
                .cmp_lte,
                .cmp_eq,
                .cmp_gte,
                .cmp_gt,
                .cmp_neq,
                .store, // a little surprising that this is technically binop, but we handle this separately anyway
                .bool_and,
                .bool_or,
                .bit_and,
                .bit_or,
                .xor,
```

So for this case, we can remove the outer switch as you suggested.

It's the other case below where we might need the outer switch for, because `emitConstant` needs to handle any `inst.Constant`, where we need to handle (nearly) all types.